st2common/st2common/metrics/base.py

```diff
@@ -0,0 +1,177 @@
+        else:
+            self._metrics.time(self.key, time_delta.total_seconds())
+
+    def get_time_delta(self):
+        """ Get current time delta.
+        """
+        return get_datetime_utc_now() - self._start_time
+
+    def __enter__(self):
+        self._start_time = get_datetime_utc_now()
+        return self
+
+    def __exit__(self, *args):
+        self.send_time()
+
+    def __call__(self, func):
+        @wraps(func)
+        def wrapper(*args, **kw):
+            with self as metrics_timer:
+                return func(*args, metrics_timer=metrics_timer, **kw)
+        return wrapper
+
+
+class Counter(object):
+    """ Timer context manager for easily sending timer statistics.
+    """
+    def __init__(self, key):
+        assert isinstance(key, str)
+        assert key
+        self.key = key
+        self._metrics = get_driver()
+
+    def __enter__(self):
+        self._metrics.inc_counter(self.key)
+        return self
+
+    def __exit__(self, *args):
+        self._metrics.dec_counter(self.key)
+
+    def __call__(self, func):
+        @wraps(func)
+        def wrapper(*args, **kw):
+            with self:
+                return func(*args, **kw)
+        return wrapper
+
+
+class CounterWithTimer(object):
+    """ Timer and counter context manager for easily sending timer statistics
+    with builtin timer.
+    """
+    def __init__(self, key):
+        assert isinstance(key, str)
+        assert key
+        self.key = key
+        self._metrics = get_driver()
+        self._start_time = None
+
+    def send_time(self, key=None):
+        """ Send current time from start time.
+        """
+        time_delta = get_datetime_utc_now() - self._start_time
+
+        if key:
+            assert isinstance(key, str)
+            self._metrics.time(key, time_delta.total_seconds())
+        else:
+            self._metrics.time("%s%s" % (self.key, METRICS_TIMER_SUFFIX),
+                               time_delta.total_seconds())
+
+    def get_time_delta(self):
+        """ Get current time delta.
+        """
+        return get_datetime_utc_now() - self._start_time
+
+    def __enter__(self):
+        self._metrics.inc_counter("%s%s" % (self.key, METRICS_COUNTER_SUFFIX))
+        self._start_time = get_datetime_utc_now()
+        return self
+
+    def __exit__(self, *args):
+        self.send_time()
+        self._metrics.dec_counter("%s_counter" % self.key)
+
+    def __call__(self, func):
+        @wraps(func)
+        def wrapper(*args, **kw):
+            with self as counter_with_timer:
+                return func(*args, metrics_counter_with_timer=counter_with_timer, **kw)
+        return wrapper
+
+
+def metrics_initialize():
+    """Initialize metrics constant
+    """
+    try:
+        _METRICS = get_plugin_instance(PLUGIN_NAMESPACE, cfg.CONF.metrics.driver)()
+    except (NoMatches, MultipleMatches, NoSuchOptError):
+        _METRICS = BaseMetricsDriver()
```

Comment:
If we implement the noop metrics driver and make it a default, we don't need this try-catch block.